mono/mini/aot-runtime.c

```diff
@@ -5167,23 +5166,32 @@ init_plt (MonoAotModule *amodule)
 	if (amodule->plt_inited)
 		return;
 
+	tramp = mono_create_specific_trampoline (amodule, MONO_TRAMPOLINE_AOT_PLT, mono_get_root_domain (), NULL);
+	tramp = mono_create_ftnptr (mono_domain_get (), tramp);
+
+	amodule_lock (amodule);
+
+	if (amodule->plt_inited) {
+		amodule_unlock (amodule);
+		return;
+	}
+
 	if (amodule->info.plt_size <= 1) {
 		amodule->plt_inited = TRUE;
+		amodule_unlock (amodule);
 		return;
 	}
 
-	tramp = mono_create_specific_trampoline (amodule, MONO_TRAMPOLINE_AOT_PLT, mono_get_root_domain (), NULL);
-
 	/*
 	 * Initialize the PLT entries in the GOT to point to the default targets.
 	 */
-
-	tramp = mono_create_ftnptr (mono_domain_get (), tramp);
 	 for (i = 1; i < amodule->info.plt_size; ++i)
 		 /* All the default entries point to the AOT trampoline */
 		 ((gpointer*)amodule->got)[amodule->info.plt_got_offset_base + i] = tramp;
 
 	amodule->plt_inited = TRUE;
```

Comment:
I think now you need a barrier before you set plt_inited to true since now we check plt_inited outside the lock sometimes.

Reply:
Fixed.